src/components/KeyboardNavigation/Provider.js

```diff
@@ -2,32 +2,95 @@ import React from "react";
 import { Provider } from "./Context";
 
+const { forEach } = React.Children;
 const ARROWUP = 38;
 const ARROWDOWN = 40;
-const SPACEBAR = 32;
-const TABKEY = 9;
-const ENTERKEY = 13;
-
-const WHITELIST = [ARROWUP, ARROWDOWN, SPACEBAR, TABKEY, ENTERKEY];
 
 export default class KeyBoardProvider extends React.Component {
   static propTypes = {
-    children: PropTypes.node.isRequired,
-    role: PropTypes.string.isRequired
+    children(props, propName, componentName) {
+      if (props[propName] === null) return new Error("Invalid Child");
+      forEach(props[propName], child => {
+        if (!child.props.value || !child.props.index) {
+          return new Error(
+            `Invalid Child \`${child}\` supplied to` +
+              ` \`${componentName}\`. Validation failed.`
+          );
+        }
+        return null;
+      });
+      return null;
+    },
+
+    role: PropTypes.string.isRequired,
+    className: PropTypes.string,
+    keywordSearch: PropTypes.bool,
+    keyBoardRef: PropTypes.func
   };
+
+  static defaultProps = {
+    children: null,
+    keyBoardRef: null,
+    className: null,
+    keywordSearch: false
+  };
+
+  static getDerivedStateFromProps(props, state) {
+    if (!props.selected || props.selected.length === 0) return null;
+
+    if (state.firstUpdate === true) {
+      const childrenArray = React.Children.toArray(props.children);
+
+      const valueOfSelected = props.selected[0];
+      const selectedChild = childrenArray.find(
+        thisArg => thisArg.props.value === valueOfSelected
+      );
+      return {
+        focused: selectedChild.props.index,
+        firstUpdate: false
+      };
+    }
+
+    return null;
+  }
+
   onClick = ({ index }) => {
     this.setState(() => ({ focused: index }));
+    this.setState({ word: "" });
   };
 
   onKeyDown = event => {
-    if (!WHITELIST.includes(event.keyCode)) return;
+    const { keywordSearch } = this.props;
+
+    // heres the logic for looking up words with in the drop down options with the keyboard
+    // it currently matches based on the text that has been typed
+    if (keywordSearch) {
+      const { word } = this.state;
+      if (event.keyCode > 48 && event.keyCode < 90) {
+        const temp = word + event.key;
+        this.setState({ word: temp }, () => this.searchValue(temp));
+      }
 
-    /* eslint-disable */
-    ![TABKEY, ENTERKEY, SPACEBAR].includes(event.keyCode) &&
+      if (event.keyCode === 32) this.setState({ word: "" });
+    }
+
+    if ([ARROWUP, ARROWDOWN].includes(event.keyCode)) {
       event.preventDefault();
-    /* eslint-enable */
+      this.cycleSelect(event.keyCode);
+    }
+  };
+
+  searchValue = string => {
+    const { children } = this.props;
 
-    this.cycleSelect(event.keyCode);
+    const childrenArray = React.Children.toArray(children);
+
+    const firstMatch = childrenArray.find(
+      thisArg =>
+        thisArg.props.children.substring(0, string.length).toLowerCase() ===
```

Comment:
Will this break if an element, rather than a string, is passed as a child?
Also, I feel like a regex may be better suited for string-matching, rather than strict equivalence, so we can treat the user's input as case insensitive.